### ISSUE TITLE
[discussion] Ability to define how to collapse dirs

### ISSUE THREAD

**Issue by yyoncho** [Alexander-Miller/treemacs#290]
For some of the languages(e. g. Java) the way the folders are grouped is defined on the language semantics. For example, I would like to group `src/main/java` `src/test/java` and then start the collapsing from that point with the same algorithm as now. I believe that it might be handy for other languages like Python as well.

**Comment by Alexander-Miller** [comment]
You'll get my usual answer to your requests: sounds good, but won't be that easy. Basically the whole collapse process is the wrong spot to implement this. Collapsing only changes the display and sets some bits of meta data, *after* the mainline display is done. To make what you ask work we need 2 things:
 * Adding more directories into the list, so just /src can become /src/main/java, /src/test/java, /src/main/resources, etc.
 * Sorting must be taken into account since at least in java src dirs are at the top.

A parallel python process, as with git, is a no-go, as it is the fetching and sorting of directory content that these python processes run in parallel to, so performance will be a concern. On the other hand `treemacs--get-dir-content` is an ideal candidate for a native module. In fact I went ahead and rewritten it in rust. Going with a straightforward re-implementation (get dirs and files, split them and sort them alphabetically) reduces the runtime by half, so that's plenty of breathing room.

In addition this sort of thing needn't be a static transformation. Eclipse lets you elevate any directory to a src dir, treemacs could do the same. Treemacs does not really have project properties yet, but the foundation has been laid. I don't know if you've taken a look at the treemacs-persist-file recently, but it's all org-mode now. So far all projects have is a `path` property, but all the code that went into this has been written with the intention of enabling custom user properties like these.

**Comment by yyoncho** [comment]
I used "collapse" since the variable handling the directory `collapsing` is `treemacs-collapse-dirs`.

I wonder whether it could be handled by the extension package, e. g. Insert extension on the top level and then hide the folder in its original place. Doing this way it will push the complexity of implementing this outside of `treemacs` core.

> So far all projects have is a path property, but all the code that went into this has been written with the intention of enabling custom user properties like these.

I think that this broke my extension code since I was using `:path` key.

**Comment by yyoncho** [comment]
Forgot to mention that this could be long-term low priority feature request.

**Comment by Alexander-Miller** [comment]
> Doing this way it will push the complexity of implementing this outside of treemacs core.

Also an interesting idea, but the performance issues would remain. Unless java's language server is able to efficiently give you a list of all source dirs or even packages, without having to dig through the entire file system first.

> I think that this broke my extension code since I was using :path key.

I'll have to remember to add a list of taken properties to the coming documentation. At any rate the :path property is not the same as a project property, though there is some small overlap. The former is a text property stored in the treemacs buffer, the latter is something I'd have to add to the treemacs-project struct.

> Forgot to mention that this could be long-term low priority feature request.

That's for the best. My plate is full enough as it is, and dir-level extensions are more difficult than I'd have thought. Since they can have any dir node as parent instead of only projects, so their paths are different than I've anticipated.

**Comment by yyoncho** [comment]
> Also an interesting idea, but the performance issues would remain. Unless java's language server is able to efficiently give you a list of all source dirs or even packages, without having to dig through the entire file system first.

Actually, this is what I was looking for - I have the source directories.

**Comment by Alexander-Miller** [comment]
Just pushed a not very tested version of `treemacs-define-directory-extension` to master. goto- and update-node work the same, but instead of the project it's the filepath of the parent that goes first in their path. That api isn't solid yet, and I've plenty of cleanup to do anyway.

**Comment by Alexander-Miller** [comment]
>  I know the source directories upfront.

Just the source directories, or could a language server provide an alternative means of fetching a project root's files instead of calling `directory-files`? In fact, the more of `treemacs--get-dir-content` can be replaced the better.

**Comment by Alexander-Miller** [comment]
I've got time now. What's the status here?

**Comment by yyoncho** [comment]
I do not think that this is high priority from lsp-mode perspective. I would trade that for https://github.com/Alexander-Miller/treemacs/issues/441

**Comment by Alexander-Miller** [comment]
Stayin alive.
